Improve support for Sony TV devices

martomo opened the pull request:
* Resolve issue with `current_app` not representing focused application when screen is turned off (by [**Picture Off** action](https://helpguide.sony.net/tv/eusltn3/v1/en-003/01-03_04.html#SID_004069)).
* Fix trailing `}` character in `current_app` when focused application has no activity.
* Added additional app constants for Sony TV system applications.
* Properly return hardware identifier for external input as `hdmi_input`. (fixes #227)

**Disclaimer:** This solution was developed on a Sony Bravia KD-55XD7005 (PKG6.6545.0255EUA) running Android 8.0.0.
Test cases written for other devices are based on output reported in relevant conversations, although I have no way to actually verify this solution works on other devices.

JeffLIrion commented:
Can you please split off the app name constants (i.e., `APP_*`) into a separate pull request. I'll merge that right away.

JeffLIrion commented on androidtv/constants.py:
```diff
@@ -28,20 +30,30 @@
 
 #: Get the HDMI input
-CMD_HDMI_INPUT = "dumpsys activity starter | grep -o 'HDMIInputService\\/HW[0-9]' -m 1 | grep -o 'HW[0-9]'"
+CMD_HDMI_INPUT_SEARCH = r" | grep -E -o '(ExternalTv|HDMI)InputService/HW[0-9]' -m 1 | grep -o 'HW[0-9]'"
```
I'd rather not have a constant that's only used once.

JeffLIrion replied:
Again, it seems Chrome on iOS failed me. I see that you also used it in a unit test. 

Still, I don't think it's necessary to have this constant. Your unit test could simply strip off "dumpsys activity starter" from the command.

martomo replied:
Nice suggestion! Instead of stripping `dumpsys activity starter`, I decided to replace `dumpsys activity starter` with the dumpsys output as specified in test case.

JeffLIrion commented on androidtv/constants.py:
```diff
@@ -28,20 +30,30 @@
 CMD_AWAKE = "dumpsys power | grep mWakefulness | grep -q Awake"
 
+#: Define current app
+CMD_DEFINE_CURRENT_APP = "CURRENT_APP=$(dumpsys window windows | grep -E 'mCurrentFocus|mFocusedApp') && " + VAR_CURRENT_APP_EXTRACTION
```
This constant seems unnecessary. Same goes for the other `CMD_DEFINE_CURRENT_APP_*` constant.

martomo replied:
In which way do you see this as potentially unnecessary?

Besides the main objective to deduplicate the code, as the majority of the complexity is already defined in `VAR_CURRENT_APP_EXTRACTION`.

The other use case I had in mind, using this module from a Python script and being able to use `CURRENT_APP` (directly) in a custom command using **adb shell** without the additional (`CMD_CURRENT_APP`) command to obtain the `current_app`.

JeffLIrion replied:
My comments from earlier were based on looking at this pull request on my phone, and the "Find in page" functionality wasn't working right.  So I do see that these constants are used and that they make the code a bit more DRY.  

My main concerns are that

1. Adding more constants makes the pull request harder to review -- it's hard to see what is actually being changed.
2. Aside from this pull request, it's easier to read the code if the full command is spelled out:

```python
# I need to check what `VAR_CURRENT_APP_EXTRACTION` does
CMD_DEFINE_CURRENT_APP = "CURRENT_APP=$(dumpsys window windows | grep -E 'mCurrentFocus|mFocusedApp') && " + VAR_CURRENT_APP_EXTRACTION

# Here I know exactly what the ADB shell command is doing
CMD_CURRENT_APP = "CURRENT_APP=$(dumpsys window windows | grep mCurrentFocus) && CURRENT_APP=${CURRENT_APP#*{* * } && CURRENT_APP=${CURRENT_APP%%/*} && echo $CURRENT_APP"
```

> The other use case I had in mind, using this module from a Python script and being able to use `CURRENT_APP` (directly) in a custom command using **adb shell** without the additional (`CMD_CURRENT_APP`) command to obtain the `current_app`.

The way I see it, anyone doing that is a power user.  I think such a person would be more likely to look through these constants and piece together what they need in order to achieve their desired outcome OR simply formulate their ADB shell command from scratch rather than to pick out constants from this module.

martomo replied:
To assist with the review of the code, below is the diff of fully written commands by changed constant:

**CMD_CURRENT_APP**
```diff
-CURRENT_APP=$(dumpsys window windows | grep mCurrentFocus) && CURRENT_APP=${CURRENT_APP#*{* * } && CURRENT_APP=${CURRENT_APP%%/*} && echo $CURRENT_APP
+CURRENT_APP=$(dumpsys window windows | grep -E 'mCurrentFocus|mFocusedApp') && CURRENT_APP=${CURRENT_APP#*ActivityRecord{* * } CURRENT_APP=${CURRENT_APP#*{* * } && CURRENT_APP=${CURRENT_APP%%/*} && CURRENT_APP=${CURRENT_APP%\\}*} && echo $CURRENT_APP
```

**CMD_CURRENT_APP_GOOGLE_TV**
```diff
-CURRENT_APP=$(dumpsys activity a . | grep -E 'mResumedActivity' | cut -d ' ' -f 8) && CURRENT_APP=${CURRENT_APP%%/*} && echo $CURRENT_APP
+CURRENT_APP=$(dumpsys activity a . | grep mResumedActivity) && CURRENT_APP=${CURRENT_APP#*ActivityRecord{* * } CURRENT_APP=${CURRENT_APP#*{* * } && CURRENT_APP=${CURRENT_APP%%/*} && CURRENT_APP=${CURRENT_APP%\\}*} && echo $CURRENT_APP
```

**CMD_HDMI_INPUT**
```diff
-dumpsys activity starter | grep -o 'HDMIInputService\\/HW[0-9]' -m 1 | grep -o 'HW[0-9]'
+dumpsys activity starter | grep -E -o '(ExternalTv|HDMI)InputService/HW[0-9]' | grep -o 'HW[0-9]'
```

**CMD_LAUNCH_APP**
```diff
-CURRENT_APP=$(dumpsys window windows | grep mCurrentFocus) && CURRENT_APP=${{CURRENT_APP#*{{* * }} && CURRENT_APP=${{CURRENT_APP%%/*}} && if [ $CURRENT_APP != '{0}' ]; then monkey -p {0} -c " + INTENT_LAUNCH + " --pct-syskeys 0 1; fi
+CURRENT_APP=$(dumpsys window windows | grep -E 'mCurrentFocus|mFocusedApp') && CURRENT_APP=${{CURRENT_APP#*ActivityRecord{{* * }} CURRENT_APP=${{CURRENT_APP#*{{* * }} && CURRENT_APP=${{CURRENT_APP%%/*}} && CURRENT_APP=${{CURRENT_APP%\\}}*}} && if [ $CURRENT_APP != '{0}' ]; then monkey -p {0} -c " + INTENT_LAUNCH + ' --pct-syskeys 0 1; fi
```

**CMD_LAUNCH_APP_GOOGLE_TV**
```diff
-CURRENT_APP=$(dumpsys activity a . | grep -E 'mResumedActivity' | cut -d ' ' -f 8) && CURRENT_APP=${{CURRENT_APP%%/*}} && if [ $CURRENT_APP != '{0}' ]; then monkey -p {0} -c " + INTENT_LAUNCH + " --pct-syskeys 0 1; fi
+CURRENT_APP=$(dumpsys activity a . | grep mResumedActivity) && CURRENT_APP=${{CURRENT_APP#*ActivityRecord{{* * }} CURRENT_APP=${{CURRENT_APP#*{{* * }} && CURRENT_APP=${{CURRENT_APP%%/*}} && CURRENT_APP=${{CURRENT_APP%\\}}*}} && if [ $CURRENT_APP != '{0}' ]; then monkey -p {0} -c " + INTENT_LAUNCH + ' --pct-syskeys 0 1; fi
```

_Note_: If you plan to compare these commands via [android.adb_command service](https://www.home-assistant.io/integrations/androidtv/#androidtvadb_command) in Home Assistant, be sure to remove any escaped characters (e.g. `\\`, `{{`, `}}`) or to populate placeholders (`{0}`).

In regard to having more constants/variables, by simply looking at the above diffs you can see there is a lot of duplicate complexity, this should simplify any future changes to `CURRENT_APP` logic, as already mentioned in https://github.com/JeffLIrion/python-androidtv/pull/233#discussion_r571491318.

JeffLIrion commented on androidtv/constants.py:
```diff
@@ -15,6 +15,8 @@
 INTENT_HOME = "android.intent.category.HOME"
 
+# Specify application identifier as CURRENT_APP from dumpsys output previously assigned to CURRENT_APP
+VAR_CURRENT_APP_EXTRACTION = 'CURRENT_APP=${CURRENT_APP#*ActivityRecord{* * } CURRENT_APP=${CURRENT_APP#*{* * } && CURRENT_APP=${CURRENT_APP%%/*} && CURRENT_APP=${CURRENT_APP%\\}*}'
```
This isn't used.

JeffLIrion replied:
Actually, I guess it is used. Chrome on my phone only found it once on this page. So maybe it's OK.

martomo replied:
Actually this is the core "logic" in regard to extracting the application identifier from the dumpsys output, either it being from `dumpsys window windows` or `dumpsys activity a .`.

It does the following operations:
1. `${CURRENT_APP#*ActivityRecord{* * }`
Attempt to extract application activity from `mResumedActivity` or `mFocusedApp`
2. `${CURRENT_APP#*{* * }`
Fallback to extract application activity from `mCurrentFocus`, in case `mFocusedApp` is not defined
3. `${CURRENT_APP%%/*}`
Remove extraneous activity details for extracted application, leaving only identifier
4. `${CURRENT_APP%\\}*}`
Delete trailing curly brace from identifier of extracted application, in case application has no activity

JeffLIrion commented on androidtv/constants.py:
```diff
@@ -28,20 +30,30 @@
+CMD_LAUNCH_APP_CONDITION = "if [ $CURRENT_APP != '{0}' ]; then monkey -p {0} -c " + INTENT_LAUNCH + ' --pct-syskeys 0 1; fi'
+
+#: Launch an app
+CMD_LAUNCH_APP = CMD_DEFINE_CURRENT_APP.replace('{', '{{').replace('}', '}}') + ' && ' + CMD_LAUNCH_APP_CONDITION
```
I know it's not DRY to copy & paste, but given that this module is simply a bunch of constants, I think it's acceptable here. Also, I think it's much more readable.

martomo replied:
I certainly agree in regard to readability, also slightly consider this to be a despicable solution, but it does escape all curly braces in the referenced constant to accompany the formatting done with this constant definition.

Although the complexity of extracting the `CURRENT_APP` variable from dumpsys output, from my experience, having multiple occurrences with duplicate logic, it's very easy to overlook certain occurrences when updating logic in regard to `CURRENT_APP`.

This is the main reason I took the DRY approach, by introducing the `CMD_DEFINE_CURRENT_APP` and `VAR_CURRENT_APP_EXTRACTION` constants, for reusability and to simplifying maintenance of `CURRENT_APP` logic.

JeffLIrion commented on androidtv/constants.py:
```diff
@@ -28,20 +30,29 @@
+CMD_DEFINE_CURRENT_APP_GOOGLE_TV = 'CURRENT_APP=$(dumpsys activity a . | grep mResumedActivity) && ' + VAR_CURRENT_APP_EXTRACTION
 
 #: Get the current app for a Google TV device
-CMD_CURRENT_APP_GOOGLE_TV = "CURRENT_APP=$(dumpsys activity a . | grep -E 'mResumedActivity' | cut -d ' ' -f 8) && CURRENT_APP=${CURRENT_APP%%/*} && echo $CURRENT_APP"
```
It looks like you changed the command here (old first, new second):

```
CURRENT_APP=$(dumpsys activity a . | grep -E 'mResumedActivity' | cut -d ' ' -f 8) ...
CURRENT_APP=$(dumpsys activity a . | grep mResumedActivity) ...
```

Why the change?  I don't own a Google TV device, so I can't test whether this works.

martomo replied:
For this particular change I looked at this [Home Assistant Community topic](https://community.home-assistant.io/t/new-chromecast-w-android-tv-integration-only-showing-as-off-or-idle/234424/30) (reference from [#207](https://github.com/JeffLIrion/python-androidtv/issues/207#issuecomment-721840720)). Indicating that both `mCurrentFocus` and `mFocusedApp` are no longer specified in the `dumpsys window windows` output.

Then I checked the alternative command (`dumpsys activity a . | grep mResumedActivity`) to obtain application activity details, which is also working on my Sony TV running Android 8.0.0, and noticed it has the same **ActivityRecord** entry as `mFocusedApp`.

```
mResumedActivity: ActivityRecord{35dbb2a u0 com.google.android.tvlauncher/.MainActivity t2719}
```
```
mFocusedApp=AppWindowToken{c791eb8 token=Token{a6bad1b ActivityRecord{35dbb2a u0 com.google.android.tvlauncher/.MainActivity t2719}}}
```

Rather than using the existing solution which relies on space (indentation) to extract the application activity details, I decided to change this approach to be more generic by looking for the **ActivityRecord** instead.

martomo commented on tests/test_constants.py:
```diff
@@ -7,12 +8,390 @@
+               keysPaused=false inHistory=true visible=false sleeping=false idle=true mStartingWindowState=STARTING_WINDOW_SHOWN
+               fullscreen=true noDisplay=false immersive=false launchMode=2
+               frozenBeforeDestroy=false forceNewConfig=false
+               mActivityType=HOME_ACTIVITY_TYPE
+               waitingVisible=false nowVisible=false lastVisibleTime=-2m58s442ms
+               resizeMode=RESIZE_MODE_UNRESIZEABLE
+               mLastReportedMultiWindowMode=false mLastReportedPictureInPictureMode=false
+              mStartActivity:
+               packageName=com.sony.dtv.tvx processName=com.sony.dtv.tvx
+               launchedFromUid=10134 launchedFromPackage=com.sony.dtv.tvx userId=0
+               app=null
+               Intent { act=android.intent.action.VIEW dat=content://android.media.tv/passthrough/com.sony.dtv.tvinput.external%2F.ExternalTvInputService%2FHDMI100004 flg=0x10400000 cmp=com.sony.dtv.tvx/.MainActivity (has extras) }
+               frontOfTask=false task=TaskRecord{9eecc8c #2830 A=com.sony.dtv.tvx U=0 StackId=1 sz=1}
+               taskAffinity=com.sony.dtv.tvx
+               realActivity=com.sony.dtv.tvx/.MainActivity
+               baseDir=/system/priv-app/Tvx/Tvx.apk
+               dataDir=/data/user/0/com.sony.dtv.tvx
+               stateNotNeeded=false componentSpecified=false mActivityType=0
+               compat=null labelRes=0x7f090048 icon=0x7f02006d theme=0x7f0b01c2
+               mLastReportedConfigurations:
+                mGlobalConfig={1.0 ?mcc?mnc [en_GB] ldltr sw540dp w960dp h540dp 320dpi lrg long hdr land television -touch -keyb/v/h dpad/v appBounds=Rect(0, 0 - 1920, 1080) s.3}
+                mOverrideConfig={0.0 ?mcc?mnc ?localeList ?layoutDir ?swdp ?wdp ?hdp ?density ?lsize ?long ?ldr ?wideColorGamut ?orien ?uimode ?night ?touch ?keyb/?/? ?nav/?}
+               CurrentConfiguration={1.0 ?mcc?mnc [en_GB] ldltr sw540dp w960dp h540dp 320dpi lrg long hdr land television -touch -keyb/v/h dpad/v appBounds=Rect(0, 0 - 1920, 1080) s.3}
+               launchFailed=false launchCount=0 lastLaunchTime=0
+               haveState=true icicle=null
+               state=INITIALIZING stopped=false delayedResume=false finishing=false
+               keysPaused=false inHistory=false visible=false sleeping=false idle=false mStartingWindowState=STARTING_WINDOW_NOT_SHOWN
+               fullscreen=true noDisplay=false immersive=false launchMode=3
+               frozenBeforeDestroy=false forceNewConfig=false
+               mActivityType=APPLICATION_ACTIVITY_TYPE
+               resizeMode=RESIZE_MODE_RESIZEABLE
+               mLastReportedMultiWindowMode=false mLastReportedPictureInPictureMode=false
+               supportsPictureInPicture=true
+               supportsPictureInPictureWhilePausing: false
+              mIntent=Intent { act=android.intent.action.VIEW dat=content://android.media.tv/passthrough/com.sony.dtv.tvinput.external/.ExternalTvInputService/HDMI100004 flg=0x10400000 cmp=com.sony.dtv.tvx/.MainActivity (has extras) }
+              mLaunchSingleTop=false mLaunchSingleInstance=true mLaunchSingleTask=false mLaunchFlags=0x10000000 mDoResume=true mAddingToTask=false
+        """
+        hdmi_input = self._hdmi_input(dumpsys_output)
+
+        self.assertEqual(hdmi_input, '')
```
Going forward I would highly recommend to enforce tests cases that actually assert commands against dumpsys output fixtures, which would for sure give me more confidence about (future) changes made to these constants.

JeffLIrion replied:
That's a good idea. I'd thought about it but never took action on it, so thanks for contributing this.

JeffLIrion commented:
I checked out your branch, rebased it onto master, and renamed a couple things.  Here's the pull request: https://github.com/JeffLIrion/python-androidtv/pull/236

In particular, this commit makes it easier to see the difference in the ADB shell commands: https://github.com/JeffLIrion/python-androidtv/pull/236/commits/abe39dec6e7a889f96abfa35295f6fd6507dfa4f

martomo commented:
> I checked out your branch, rebased it onto master, and renamed a couple things. Here's the pull request: #236

I would have been open to any suggestions for renaming and having them addressed in this pull request.

> In particular, this commit makes it easier to see the difference in the ADB shell commands: [abe39de](https://github.com/JeffLIrion/python-androidtv/commit/abe39dec6e7a889f96abfa35295f6fd6507dfa4f)

I guess my https://github.com/JeffLIrion/python-androidtv/pull/233#discussion_r571585657 was not helpful with reviewing the difference? I'm glad you managed in the end and really like your approach to outputting fully written commands in tests (#235) for reviewing purposes.

JeffLIrion commented:
> I guess my [#233 (comment)](https://github.com/JeffLIrion/python-androidtv/pull/233#discussion_r571585657) was not helpful with reviewing the difference? I'm glad you managed in the end and really like your approach to outputting fully written commands in tests (#235) for reviewing purposes.

Your comment sparked my idea. By including a test for it, I can guarantee that the old and the new commands are both being reported correctly. 

Anyways, I just figured it was easier to make some of the renaming changes myself (and also update the commands test so that it passes) rather than to comment the changes I wanted on your pull request and have you make the changes.
